_posts/2016-11-02-improving-page-speed-with-graphql.markdown

```diff
@@ -0,0 +1,113 @@
+})
+
+# Define api with JoiQL like this:
+const api = joiql({
+  query: {
+    article: Article
+  }
+})
+
+# Resolve the request using a Koa 2 style middleware pattern:
+api.use((ctx, next) => {
+  return new Promise (resolve, reject) ->
+    # Method that fetches an article based on the query
+    findArticle(ctx.req.query.article.args), (err, results) ->
+      ctx.res.article = results.article
+      next()
+      resolve()
+})
+
+# Finally, mount our schema to express:
+app.use('/graphql', graphqlHTTP({
+  schema: api.schema,
+  graphiql: true
+}))
+app.listen(3000, () => console.log('listening on 3000'))
+
+```
+
+You can see how simple it becomes to convert apps that already use Joi with JoiQL. It's less JSON blobs to handle and we also get a nice Koa 2 style middleware pattern that lets us hook into a downstream/upstream flow. This could be useful later on when if we were to get really granular with measuring speed and decide to track the speed of a request manually. We could do something like this:
+
+```
+api.use((ctx, next) => {
+  startTIme = Date.now()
+  next()
+  console.log(Date.now() - startTime)
+
+api.use((ctx, next) => {
+  # fetch content here
+  next() #we get bumped back up after this
+})
+```
+
+## Results
+
+The two features of GraphQL that have been helpful for reducing page speed are:
+1. Reduced payload because you only request the data you need
+2. Multiple fetches can be coalesced into a single request
+
+Not surprisingly, decreased payload and coalesced requests are the same two features [Orta](http://twitter.com/orta) describes in part of his post on the killer features of [GraphQL for Mobile](http://artsy.github.io/blog/2016/06/19/graphql-for-mobile).
+
+### Reduced Payload
+
+Reduced payload turned out to be one of the biggest factors in reducing speed. Just after October 27 we switched over to using GraphQL on our [mobile articles landing page](http://m.artsy.net/articles).
+
+{% expanded_img /images/2016-11-02-improving-page-speed-with-graphql/download_time.png %}
+
+Since we were previously getting back collection of articles with all of its content, rendering a simple page with a list of articles meant the payload size was unnecessarily huge. Now, we only request a few things like the title and image without having to do any extra work. This is especially important for mobile devices where we encounter slow network speeds!
+
+### Multiple fetches
+
+Although we haven't made use of this in production yet, I anticipate that the ability to coalesce requests will be significant. For example, to render an article page, there are up to five fetches that can be requested by our front-end app with a normal REST API. Now, we can combine the requests to a single fetch, which then moves the responsibility of fanning out database queries to the API. If we aggregate the request and cache the response using a client like [Lokka](https://github.com/kadirahq/lokka), our future looks bright.
```

Comment:
This line is a little unclear to me, I'll propose a revision that I hope is helpful.

> Now, we can combine the requests to a single fetch, which then moves the responsibility of fanning out database queries to the API. 

Now, we can combine the requests to a single fetch, which moves the responsibility of coalescing requests from aggregating slow HTTP requests to fanning out fast database queries.